gdbm: restrict patch to 1.11:1.18, add nls variant

iarspider opened the pull request:
File to patch missing in 1.10. Also restruct `readline` dependency to 1.13 and newer.

iarspider committed:
gdbm: restrict patch to 1.11:1.18, add nls variant

haampie commented on var/spack/repos/builtin/packages/gdbm/package.py:
```diff
@@ -29,15 +29,17 @@ class Gdbm(AutotoolsPackage, GNUMirrorPackage):
-    depends_on("readline")
+    variant('nls', default=True, description="Use Native Language Support")
+
+    depends_on("readline", when="+nls")
```
This looks really weird to me. What does translation have to do with command line stuff? Don't you mean to add a dependency on gettext when +nls?

iarspider replied:
@haampie good catch.

iarspider commented:
I have no clue why the style test fails. The reported error is quite useless:
```
Run share/spack/qa/run-style-tests
  share/spack/qa/run-style-tests
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.10.5/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.10.5/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.5/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.5/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.10.5/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.10.5/x64/lib
==> Running style checks on spack
  selected: isort, mypy, black, flake8
==> Modified files
  var/spack/repos/builtin/packages/gdbm/package.py
==> Error: cannot bootstrap any of the isort executables from spec "py-isort@4.3.5: %gcc target=x86_64" due to the following failures:
==> Bootstrapping py-isort from pre-built binaries
github-actions-v0.2 raised FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/spack/spack/share/spack/bootstrap/github-actions-v0.2/py-isort.json'
==> Bootstrapping py-isort@4.3.5: %gcc target=x86_64 from sources
github-actions-v0.1 raised ValueError: source is not trusted
==> Bootstrapping clingo from pre-built binaries
spack-install raised SpecParseError: Unexpected token: ':'
==> Fetching https://mirror.spack.io/bootstrap/github-actions/v0.2/build_cache/linux-centos7-x86_64-gcc-10.2.1-clingo-bootstrap-spack-2l45t4kw3cqqwj6vbxhfwhzlo6b3q2p4.spec.json
Run `spack --debug ...` for more detailed errors
==> Fetching https://mirror.spack.io/bootstrap/github-actions/v0.2/build_cache/linux-centos7-x86_64/gcc-10.2.1/clingo-bootstrap-spack/linux-centos7-x86_64-gcc-10.2.1-clingo-bootstrap-spack-2l45t4kw3cqqwj6vbxhfwhzlo6b3q2p4.spack
==> Installing "clingo-bootstrap@spack%gcc@10.2.1~docs~ipo+python+static_libstdcpp build_type=Release arch=linux-centos7-x86_64" from a buildcache
Error: Process completed with exit code 3.
```

When running `spack style` locally, the only reported error is in spack core:
```
==> Running style checks on spack
  selected: isort, mypy, black, flake8
==> Modified files
  var/spack/repos/builtin/packages/gdbm/package.py
==> Running isort checks
  isort checks were clean
==> Running mypy checks
lib/spack/spack/operating_systems/windows_os.py:40: error: No overload variant of "check_output" matches argument types "List[str]", "Dict[str, str]"  [call-overload]
lib/spack/spack/operating_systems/windows_os.py:40: note: Possible overload variants:
lib/spack/spack/operating_systems/windows_os.py:40: note:     def check_output(args: Union[Union[bytes, str], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., text: Literal[True]) -> str
lib/spack/spack/operating_systems/windows_os.py:40: note:     def check_output(args: Union[Union[bytes, str], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: str, errors: Optional[str] = ..., text: Optional[bool] = ...) -> str
lib/spack/spack/operating_systems/windows_os.py:40: note:     def check_output(args: Union[Union[bytes, str], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: Optional[str] = ..., errors: str, text: Optional[bool] = ...) -> str
lib/spack/spack/operating_systems/windows_os.py:40: note:     def check_output(args: Union[Union[bytes, str], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., *, universal_newlines: Literal[True], startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., text: Optional[bool] = ...) -> str
lib/spack/spack/operating_systems/windows_os.py:40: note:     def check_output(args: Union[Union[bytes, str], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: Literal[False] = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: None = ..., errors: None = ..., text: Optional[Literal[False]] = ...) -> bytes
lib/spack/spack/operating_systems/windows_os.py:40: note:     def check_output(args: Union[Union[bytes, str], Sequence[Union[str, bytes, PathLike[str], PathLike[bytes]]]], bufsize: int = ..., executable: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., stdin: Union[None, int, IO[Any]] = ..., stderr: Union[None, int, IO[Any]] = ..., preexec_fn: Optional[Callable[[], Any]] = ..., close_fds: bool = ..., shell: bool = ..., cwd: Optional[Union[str, bytes, PathLike[str], PathLike[bytes]]] = ..., env: Optional[Union[Mapping[bytes, Union[str, bytes, PathLike[str], PathLike[bytes]]], Mapping[str, Union[str, bytes, PathLike[str], PathLike[bytes]]]]] = ..., universal_newlines: bool = ..., startupinfo: Any = ..., creationflags: int = ..., restore_signals: bool = ..., start_new_session: bool = ..., pass_fds: Any = ..., *, timeout: Optional[float] = ..., input: Optional[Union[bytes, str]] = ..., encoding: Optional[str] = ..., errors: Optional[str] = ..., text: Optional[bool] = ...) -> Any
Found 1 error in 1 file (checked 556 source files)
  mypy found errors
==> Running black checks
All done! ✨ � ✨
1 file would be left unchanged.
  black checks were clean
==> Running flake8 checks
  flake8 checks were clean
==> Error: spack style found errors
```

iarspider commented:
The problem with the pipeline is: perl build fails because of `undefined reference to 'libintl_textdomain'`. There is an open ticket in Perl bugtracker https://github.com/Perl/perl5/issues/18467, in Spack bugtracker https://github.com/spack/spack/issues/21744 . There is a possible fix here: https://github.com/buildroot/buildroot/commit/e0a291244002ab19f8fc8a0adb215b51888eb5aa . Should I apply the fix to Spack recipe?

haampie commented:
Not sure if that fix is correct. My guess is that Spack's implicit `-I` flags influence Perl's configure script, making it believe that certain things are in glibc even though they're coming from gettext (and require linking). In that case adding `-lintl` would be symptom relief...

iarspider commented:
@spackbot fix style

Commit:
Fix the new nls variant to not break the build of perl and others

gdbm is happy to use the gettext part of GLibc and enables it by
default.

Not introducing gettext into the build avoids breaking the build
of perl and possibly a number of other packages using gdbm.

This means that the new nls variant allows to disable NLS in case
it is not wanted, it is enabled by default, so the default is not
changed by this PR.

PS: I rebased the initial commits of this PR to develop and squashed them.

bernhardkaindl reviewed (dismissed):
@iarspider: gdbm's configure enabled NLS by default already:
```py
spack-stage-gdbm-1.20-jrtkagtq2vmk4focstlpqc7ul7rzof6p/spack-src/configure --help | grep nls
  --disable-nls           do not use Native Language Support
```
And gdbm prefers using GNU gettext from GLibc, not from the separate library:
```py
checking for GNU gettext in libc... yes
checking whether to use NLS... yes
checking where the gettext function comes from... libc
```
This means that it is better to not add the `gettext` package to the build because it's:
- Not used by `gdbm`
- It's new presence can only break dependents of `gdbm` like `perl`.

The resulting purpose of your that variant is to allow to disable NLS in `gdbm` when not wanted.
Maybe this could help the build of gdbm on some strange systems without a fully-working glibc.

Because you didn't explain what your motivation to add the variant was, for now I am assuming this.

I rebased your commits to the current develop, squashed them and pushed them with my fixup including explanatory comments to avoid more confusion in the future.

iarspider commented:
@spackbot run pipeline

iarspider commented:
@bernhardkaindl thanks for fixing this PR. I will check why I needed `gettext` dependency (I clearly don't need it now).

iarspider committed:
Merge branch 'develop' into patch-3

iarspider committed:
Merge branch 'develop' into patch-3

bernhardkaindl commented:
@iarspider Can you check if #34286 fixes your `perl` build problem?

The bug reports you refer to are quite old, what is your exact environment and what is your full issue description?

Can you describe in which environement (host OS, etc) your problem happen? (or e.g. open a bug report for your issue with all details)?

Unless you or gdbm need to use gettext, I'd prefer if we could just disable NLS and not even make it a variant.

A very common issue with packages which enable gettext is that they are apparently only built on systems using `glibc` as the system libc, and `glibc` includes e.g. the functions which gettext provides in `libintl` and glibc also provides `/usr/include/libintl.h`. From Rocky Linux 8:

```py
$ rpm -qf /usr/include/libintl.h 
glibc-headers-2.28-216.el8.aarch64
```

When a user then tries to build such package on a non-glibc system, they might be missing `libintl.h` or the `libintl` functions are not found because they are not included in the other system libc variants, e.g. such as `musl` of Alpine Linux, or on Darwin(MacOSX). Some spack packages addressed missing `libintl` functions by adding
```py
        depends_on("gettext")
```
and sometimes even need to add:
```py
            env.append_flags("LDFLAGS", "-lintl")
```
or
```py
            args.append("-DCMAKE_SHARED_LINKER_FLAGS=-lintl")
```
But adding `-lintl` should be done only if `gettext` provides `libintl`, which is not the case when spack uses the system's `gettext` after `spack external find gettext` on glibc systems because these don't have a `libintl` provided by their system libs (since the functions are already in glibc).

glibc really mad a mess there, sigh, but that ship has sailed a loooong time ago. Here is what such packages should do then:
```py
        # Append -lintl if building against a gettext which provides a separate libintl:
        # This prevents build failures when `spack external find gettext` was used on glibc systems.
        # setup spack to use the system-provided gettext and on glibc systems libintl is "in" glibc:
        if any("libintl." in filename.split("/")[-1] for filename in self.spec["gettext"].libs):
            env.append_flags("LDFLAGS", "-lintl")
```
See #34383 if you want to read up more on that.
